lib/charms/hydra/v0/oauth.py

```diff
@@ -0,0 +1,610 @@
+        self._client_config = client_config
+        events = self._charm.on[relation_name]
+        self.framework.observe(events.relation_created, self._on_relation_created_event)
+        self.framework.observe(events.relation_changed, self._on_relation_changed_event)
+
+    def _on_relation_created_event(self, event: RelationCreatedEvent) -> None:
+        try:
+            self._update_relation_data(self._client_config, event.relation.id)
+        except ClientConfigError as e:
+            self.on.invalid_client_config.emit(e.args[0])
+
+    def _on_relation_changed_event(self, event: RelationChangedEvent) -> None:
+        if not self.model.unit.is_leader():
+            return
+
+        data = _load_data(event.relation.data[event.app], OAUTH_PROVIDER_JSON_SCHEMA)
+
+        client_id = data.get("client_id")
+        client_secret_id = data.get("client_secret_id")
+        if not client_id or not client_secret_id:
+            # The client credentials are not ready yet, so we do nothing
+            # This could mean that the client credentials were removed from the databag,
+            # but we don't allow that (for now), so we don't have to check for it.
+            return
+
+        self.on.oauth_info_changed.emit(client_id, client_secret_id)
+
+    def _update_relation_data(
+        self, client_config: Optional[ClientConfig], relation_id: Optional[int] = None
+    ) -> None:
+        if not self.model.unit.is_leader() or not client_config:
+            return
+
+        if not isinstance(client_config, ClientConfig):
+            raise ValueError(f"Unexpected client_config type: {type(client_config)}")
+
+        client_config.validate()
+
+        try:
+            relation = self.model.get_relation(
+                relation_name=self._relation_name, relation_id=relation_id
+            )
+        except TooManyRelatedAppsError:
+            raise RuntimeError("More than one relations are defined. Please provide a relation_id")
+
+        if not relation:
+            return
+
+        data = _dump_data(client_config.to_dict(), OAUTH_REQUIRER_JSON_SCHEMA)
+        relation.data[self.model.app].update(data)
+
+    def get_provider_info(self) -> Optional[Dict]:
+        """Get the provider information from the databag."""
+        if len(self.model.relations) == 0:
+            return None
+        relation = self.model.get_relation(self._relation_name)
+        if not relation:
+            return None
+
+        data = _load_data(relation.data[relation.app], OAUTH_PROVIDER_JSON_SCHEMA)
+        data.pop("client_id", None)
+        data.pop("client_secret_id", None)
+        return data
+
+    def get_client_secret(self, client_secret_id: str) -> Secret:
+        """Get the client_secret."""
+        client_secret = self.model.get_secret(id=client_secret_id)
+        return client_secret
+
+    def update_client_config(
+        self, client_config: ClientConfig, relation_id: Optional[int] = None
```

Comment:
In which cases will the user of this method know the relation id?

Reply:
Also, the `client_config` was provided in the init, why re-set it here?

Reply:
> In which cases will the user of this method know the relation id?

The user(charm) may know the `relation_id` from an event. Tbh I don't really like having the relation_id as an argument like this, but the juju/ops framework primitives around getting relations are a little weird and as far as I can tell there is no clean way of enforcing that only on relation will be created. Having more than one relations from a single oauth requirer relation does not really make sense. 

I would be OK with making the assumption in the code that there will always be at most 1 relation available (thus we would not need the relation_id to get the relation) and just having a warning in the docs.

> Also, the `client_config` was provided in the init, why re-set it here?

One of the client configuration needed is the `redirect_uri`. IMO most of the time the requirer will not know that value when initializing the `OAuthRequirer` class as that would be provided to the requirer later either from the ingress relation or user configuration. This means that the charm will need to call this function in order to update the client_config it provided.